test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java

```diff
@@ -2110,24 +2115,23 @@ public boolean test(NodeAndClient nodeAndClient) {
         }
     }
 
-    private static final class DataOrMasterNodePredicate implements Predicate<NodeAndClient> {
-        @Override
-        public boolean test(NodeAndClient nodeAndClient) {
-            return DiscoveryNode.isDataNode(nodeAndClient.node.settings()) ||
-                DiscoveryNode.isMasterNode(nodeAndClient.node.settings());
-        }
-    }
-
     private static final class MasterNodePredicate implements Predicate<NodeAndClient> {
-        private final String masterNodeName;
+        private final Optional<String> masterNodeName;
```

Comment:
I think I preferred the `DataOrMasterNodePredicate`.

Reply:
I've pushed further on the refactoring (see aa37d571490). Let me know if you like this more.